intelmq/bots/outputs/misp/output_api.py

```diff
@@ -0,0 +1,181 @@
+"""Connect to a MISP instance and add event as MISPObject if not there already.
+
+SPDX-FileCopyrightText: 2020 Intevation GmbH <https://intevation.de>
+SPDX-License-Identifier: AGPL-3.0-or-later
+
+Funding: of initial version by SUNET
+Author(s):
+  * Bernhard Reiter <bernhard@intevation.de>
+
+TODO: A shortened copy of this documentation is kept at `docs/Bots.md`, please
+      keep it current, when changing something.
+
+Parameters:
+  - add_feed_provider_as_tag: bool (use true when in doubt)
+  - misp_additional_correlation_fields: list of fields for which
+        the correlation flags will be enabled (in addition to those which are
+        in significant_fields)
+  - misp_additional_tags: list of tags to set not be searched for
+        when looking for duplicates
+  - misp_key: str, API key for accessing MISP
+  - misp_publish: bool, if a new MISP event should be set to "publish".
+        Expert setting as MISP may really make it "public"!
+        (Use false when in doubt.)
+  - misp_tag_for_bot: str, used to mark MISP events
+  - misp_to_ids_fields: list of fields for which the to_ids flags will be set
+  - misp_url: str, URL of the MISP server
+  - significant_fields: list of intelmq field names
+
+The significant field values will be searched for in all MISP attribute values
+and if all values are found in the same MISP event, no new MISP event
+will be created.
+(The reason that all values are matched without considering the
+attribute type is a technical limitation of the
+search functionality exposed by the MISP/pymisp 2.4.120 API.)
+
+If a new MISP event is inserted the `significant_fields` and the
+`misp_additional_correlation_fields` will be the attributes
+where correlation is enabled.
+
+Make sure to build the IntelMQ Botnet in a way the rate of incoming
+events is what MISP can handle, as IntelMQ can process many more events faster
+than MISP (which is by design as MISP is for manual handling).
+Also remove the fields of the IntelMQ events with an expert bot
+that you do not want to be inserted into MISP.
+
+Example (of some parameters in JSON)::
+
+    "add_feed_provider_as_tag": true,
+    "misp_additional_correlation_fields": ["source.asn"],
+    "misp_additional_tags": ["OSINT", "osint:certainty==\"90\""],
+    "misp_publish": false,
+    "misp_to_ids_fields": ["source.fqdn", "source.reverse_dns"],
+    "significant_fields": ["source.fqdn", "source.reverse_dns"],
+
+
+Originally developed with pymisp v2.4.120 (which needs python v>=3.6).
+"""
+from intelmq.lib.bot import OutputBot
+from intelmq.lib.exceptions import MissingDependencyError
+
+try:
+    import pymisp
+except ImportError:
+    pymisp = None
+
+
+class MISPAPIOutputBot(OutputBot):
+    is_multithreadable = False
```

Comment:
This was copied from `output_feed.py` and it seems right because the speed of MISP supposedly is smaller than the speed of IntelMQ. It may or may not work with multithreading, testing was a bit out of scope. Pymisp docs don't say much multi-threading either. So a single thread is probably easiest to understand. :)